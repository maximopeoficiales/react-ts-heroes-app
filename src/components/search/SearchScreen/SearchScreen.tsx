import queryString from "query-string";
import { useMemo } from "react";
import { RouteComponentProps, useLocation } from "react-router-dom";
import { heroes } from "../../../data/heroes";
import { useForm } from "../../../hooks/useForm";
import { getHeroSearch } from "../../../selectors/getHeroSearch";
import HeroCard from "../../heroes/HeroCard/HeroCard";
import Container from "../../utils/Container/Container";

interface MyProps extends RouteComponentProps<any> {}
const defaultProps = {};
const SearchScreen = (props: MyProps) => {
  props = { ...defaultProps, ...props };
  const location = useLocation();
  const { q = "" }: any = queryString.parse(location.search);

  const { history } = props;

  const { formValues, handlerChange } = useForm<{ search: string }>({
    search: q,
  });

  const { search } = formValues;
  const heroesFiltered = useMemo(() => getHeroSearch(q), [q]);

  const handlerSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    history.push(`?q=${search}`);
  };
  console.log(heroes.length);

  return (
    <div data-testid="SearchScreen" className="p-4">
      <div className="grid grid-cols-1 md:grid-cols-3">
        <div className="text-white">
          <Container>
            <form onSubmit={handlerSubmit}>
              <div className="flex flex-col my-2">
                <label htmlFor="search" className="text-base font-bold">
                  Busca un Heroe
                </label>
                <input
                  type="text"
                  className="p-2 my-2 border-none rounded-lg outline-none bg-heroBlack text-heroWhite"
                  name="search"
                  id="search"
                  value={search}
                  autoComplete="off"
                  onChange={handlerChange}
                  placeholder="Ingresa una busqueda"
                />
                <button
                  type="submit"
                  className="w-full px-3 py-2 my-2 font-bold duration-75 ease-out delay-75 transform scale-100 bg-green-600 rounded-full active:scale-75 active:bg-green-800 "
                >
                  Buscar
                </button>
              </div>
            </form>
          </Container>
        </div>
        <div className="col-span-2 p-2 mx-4">
          {heroesFiltered.length !== 0 ? (
            <>
              <h2 className="my-4 text-4xl font-bold text-center">
                {q === "" ? `All Heroes` : `Results with ${q}`}
              </h2>
              <div className="grid grid-cols-1 grid-rows-1 gap-6 md:grid-cols-2 lg:grid-cols-2">
                {heroesFiltered.map((hero) => (
                  <HeroCard key={hero.id} heroe={hero} />
                ))}
              </div>
            </>
          ) : (
            <h2 className="my-4 text-4xl font-bold text-center">
              There are no heroes with "{q}"
            </h2>
          )}
        </div>
      </div>
    </div>
  );
};
export default SearchScreen;
