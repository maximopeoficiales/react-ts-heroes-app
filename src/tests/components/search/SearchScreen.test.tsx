import { mount } from "enzyme";
import React from "react";
import { MemoryRouter, Route } from "react-router";
import SearchScreen from "../../../components/search/SearchScreen/SearchScreen";

describe("<SearchScreen />", () => {
  test("debe de mostrarse correctamente con valores por defecto", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("h2").text().trim()).toEqual("All Heroes");
  });

  test("debe de mostrar batman en el input con el valor de query string", () => {
    let query = "batman";
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=" + query]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper.find("input").prop("value")).toBe(query);
  });

  test("debe de mostrar un mensaje como titulo si no encuentra el heroe", () => {
    let query = "sdfasffd";
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=" + query]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper.find("h2").text().trim()).toEqual(
      `There are no heroes with "${query}"`
    );
  });

  test("debe llamar el push del history", () => {
    const history = {
      push: jest.fn(),
      replace: jest.fn(),
      location: {} as any,
      listen: jest.fn(),
      createHref: jest.fn(),
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route
          path="/search"
          component={() => <SearchScreen history={history as any} />}
        />
      </MemoryRouter>
    );

    let value = "batman";
    wrapper
      .find("input")
      .simulate("change", { target: { name: "search", value } });

    const formSubmit: any = wrapper.find("form").prop("onSubmit");
    formSubmit({ preventDefault() {} });

    expect(history.push).toHaveBeenCalledWith("?q=" + value);
  });
});
