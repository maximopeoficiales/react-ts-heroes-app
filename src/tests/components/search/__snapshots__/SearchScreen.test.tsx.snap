// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchScreen /> debe de mostrarse correctamente con valores por defecto 1`] = `
<div
  className="p-4"
  data-testid="SearchScreen"
>
  <div
    className="grid grid-cols-1 md:grid-cols-3"
  >
    <div
      className="text-white"
    >
      <div
        className="w-11/12 mx-auto my-8 lg:w-9/12"
      >
        <div
          className="p-5 rounded-xl lg:p-14 bg-heroBlackSecondary"
        >
          <form
            onSubmit={[Function]}
          >
            <div
              className="flex flex-col my-2"
            >
              <label
                className="text-base font-bold"
                htmlFor="search"
              >
                Busca un Heroe
              </label>
              <input
                autoComplete="off"
                className="p-2 my-2 border-none rounded-lg outline-none bg-heroBlack text-heroWhite"
                id="search"
                name="search"
                onChange={[Function]}
                placeholder="Ingresa una busqueda"
                type="text"
                value=""
              />
              <button
                className="w-full px-3 py-2 my-2 font-bold duration-75 ease-out delay-75 transform scale-100 bg-green-600 rounded-full active:scale-75 active:bg-green-800 "
                type="submit"
              >
                Buscar
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
    <div
      className="col-span-2 p-2 mx-4"
    >
      <h2
        className="my-4 text-4xl font-bold text-center"
      >
        All Heroes
      </h2>
      <div
        className="grid grid-cols-1 grid-rows-1 gap-6 md:grid-cols-2 lg:grid-cols-2"
      >
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/dc-batman"
          onClick={[Function]}
        >
          <img
            alt="dc-batman"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/dc-batman.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Batman
          </div>
        </a>
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/dc-superman"
          onClick={[Function]}
        >
          <img
            alt="dc-superman"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/dc-superman.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Superman
          </div>
        </a>
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/dc-flash"
          onClick={[Function]}
        >
          <img
            alt="dc-flash"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/dc-flash.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Flash
          </div>
        </a>
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/dc-green"
          onClick={[Function]}
        >
          <img
            alt="dc-green"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/dc-green.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Green Lantern
          </div>
        </a>
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/dc-arrow"
          onClick={[Function]}
        >
          <img
            alt="dc-arrow"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/dc-arrow.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Green Arrow
          </div>
        </a>
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/dc-wonder"
          onClick={[Function]}
        >
          <img
            alt="dc-wonder"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/dc-wonder.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Wonder Woman
          </div>
        </a>
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/dc-martian"
          onClick={[Function]}
        >
          <img
            alt="dc-martian"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/dc-martian.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Martian Manhunter
          </div>
        </a>
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/dc-robin"
          onClick={[Function]}
        >
          <img
            alt="dc-robin"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/dc-robin.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Robin/Nightwing
          </div>
        </a>
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/dc-blue"
          onClick={[Function]}
        >
          <img
            alt="dc-blue"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/dc-blue.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Blue Beetle
          </div>
        </a>
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/dc-black"
          onClick={[Function]}
        >
          <img
            alt="dc-black"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/dc-black.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Black Canary
          </div>
        </a>
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/marvel-spider"
          onClick={[Function]}
        >
          <img
            alt="marvel-spider"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/marvel-spider.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Spider Man
          </div>
        </a>
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/marvel-captain"
          onClick={[Function]}
        >
          <img
            alt="marvel-captain"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/marvel-captain.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Captain America
          </div>
        </a>
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/marvel-iron"
          onClick={[Function]}
        >
          <img
            alt="marvel-iron"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/marvel-iron.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Iron Man
          </div>
        </a>
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/marvel-thor"
          onClick={[Function]}
        >
          <img
            alt="marvel-thor"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/marvel-thor.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Thor
          </div>
        </a>
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/marvel-hulk"
          onClick={[Function]}
        >
          <img
            alt="marvel-hulk"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/marvel-hulk.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Hulk
          </div>
        </a>
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/marvel-wolverine"
          onClick={[Function]}
        >
          <img
            alt="marvel-wolverine"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/marvel-wolverine.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Wolverine
          </div>
        </a>
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/marvel-daredevil"
          onClick={[Function]}
        >
          <img
            alt="marvel-daredevil"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/marvel-daredevil.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Daredevil
          </div>
        </a>
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/marvel-hawkeye"
          onClick={[Function]}
        >
          <img
            alt="marvel-hawkeye"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/marvel-hawkeye.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Hawkeye
          </div>
        </a>
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/marvel-cyclops"
          onClick={[Function]}
        >
          <img
            alt="marvel-cyclops"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/marvel-cyclops.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Cyclops
          </div>
        </a>
        <a
          className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
          data-testid="HeroCard"
          href="/hero/marvel-silver"
          onClick={[Function]}
        >
          <img
            alt="marvel-silver"
            className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
            loading="lazy"
            src="/assets/heroes/marvel-silver.jpg"
          />
          <div
            className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
          >
            Silver Surfer
          </div>
        </a>
      </div>
    </div>
  </div>
</div>
`;
